etl-pipeline/mappings/oclc_bib.py

```diff
@@ -2,130 +2,189 @@
+    edition_cluster = oclc_bib.get('editionCluster', {}).get('id')
+
+    return [
+        f"{id}|{id_type}" 
+        for id, id_type in [(oclc_number, 'oclc'), (owi_number, 'owi'), (edition_cluster, 'oec')] 
+        if id is not None
+    ]
+
+
+def _get_publishers(oclc_bib):
+    publishers = oclc_bib.get('publishers')
+
+    if not publishers:
+        return None
     
-    def _get_contributors(self, creators):
-        if not creators:
-            return None
-
-        return list(
-            filter(
-                lambda creator: not creator.get('isPrimary', False) and not self._is_author(creator), 
-                creators
-            )
-        )
+    return [
+        f"{publisher_name}||"
+        for publisher in publishers
+        if (publisher_name := publisher.get('publisherName', {}).get('text'))
+    ]
+
+
+def _get_dates(oclc_bib):
+    dates = oclc_bib.get('date')
+
+    if not dates:
+        return None
     
-    def _is_author(self, creator):
-        for role in set(map(lambda relator: relator.get('term', '').lower(), creator.get('relators', []))):
-            if 'author' in role.lower() or 'writer' in role.lower():
-                return True
-        
-        return False
+    publication_date = dates.get('publicationDate')
+
+    if not publication_date:
+        return None
     
-    def _map_subjects(self, oclc_bib) -> list[str]:
-        return [
-            f"{subject_name}||{subject.get('vocabulary', '')}" 
-            for subject in oclc_bib.get('subjects', [])
-            if (subject_name := subject.get('subjectName', {}).get('text'))
-        ]
+    return [f'{publication_date}|publication_date']
+
+
+def _get_languages(oclc_bib):
+    languages = oclc_bib.get('language')
+
+    if not languages:
+        return None
     
-    def _map_authors(self, authors) -> Optional[list[str]]:
-        if not authors:
-            return None
-        
-        return [
-            f'{author_name}|||true' 
-            for author in authors 
-            if (author_name := self._get_contributor_name(author))
-        ]
+    item_language = languages.get('itemLanguage')
+
+    if not item_language:
+        return None
     
-    def _map_contributors(self, contributors) -> Optional[list[str]]:
-        if not contributors:
-            return None
-        
-        return [
-            f"{contributor_name}|||{', '.join(list(map(lambda relator: relator.get('term', ''), contributor.get('relators', []))))}"
-            for contributor in contributors 
-            if (contributor_name := self._get_contributor_name(contributor))
-        ]
+    return [f'||{item_language}']
+
+
+def _get_parts(oclc_bib):
+    digital_access_locations = oclc_bib.get('digitalAccessAndLocations')
+
+    if not digital_access_locations:
+        return None
+    
+    return [
+        str(Part(index=1, url=uri, source='oclc', file_type='text/html', flags=str(FileFlags(embed=True))))
```

Comment:
Should these parts all have different indices?

Reply:
i think so for the post part - they are all links to the same bib just maybe from different sources.